File: openai-agents/tour-of-agents/app/durable_agent.py
import restate

from datetime import timedelta
from agents import (
    Agent,
    RunConfig,
    RunContextWrapper,
    ModelSettings,
)
from restate import Service, VirtualObject
from restate.ext.openai import restate_context, DurableRunner, durable_function_tool

from app.utils.models import WeatherPrompt, WeatherRequest, WeatherResponse
from app.utils.utils import fetch_weather


@durable_function_tool
async def get_weather(city: WeatherRequest) -> WeatherResponse:
    """Get the current weather for a given city."""
    return await restate_context().run_typed("get weather", fetch_weather, req=city)


agent = Agent(
    name="WeatherAgent",
    instructions="You are a helpful agent that provides weather updates.",
    tools=[get_weather],
)

agent_service = restate.Service("WeatherAgent")


@agent_service.handler()
async def run(_ctx: restate.Context, req: WeatherPrompt) -> str:
    result = await DurableRunner.run(agent, req.message)
    return result.final_output


if __name__ == "__main__":
    import hypercorn
    import asyncio

    app = restate.app(services=[agent_service])
    conf = hypercorn.Config()
    conf.bind = ["0.0.0.0:9080"]
    asyncio.run(hypercorn.asyncio.serve(app, conf))

File: vercel-ai/tour-of-agents/src/durable-agent.ts
import * as restate from "@restatedev/restate-sdk";
import { openai } from "@ai-sdk/openai";
import { generateText, stepCountIs, tool, wrapLanguageModel } from "ai";
import { z } from "zod";
import { fetchWeather } from "./utils/utils";
import { durableCalls } from "@restatedev/vercel-ai-middleware";
import {WeatherPrompt, WeatherPromptSchema} from "./utils/types";
const schema = restate.serde.schema;

const run = async (ctx: restate.Context, { prompt }: WeatherPrompt) => {
  const model = wrapLanguageModel({
    model: openai("gpt-4o"),
    middleware: durableCalls(ctx, { maxRetryAttempts: 3 }),
  });

  const { text } = await generateText({
    model,
    system: "You are a helpful agent that provides weather updates.",
    prompt,
    tools: {
      getWeather: tool({
        description: "Get the current weather for a given city.",
        inputSchema: z.object({ city: z.string() }),
        execute: async ({ city }) =>
          ctx.run("get weather", () => fetchWeather(city)),
      }),
    },
    stopWhen: [stepCountIs(5)],
    providerOptions: { openai: { parallelToolCalls: false } },
  });

  return text;
};

const agent = restate.service({
  name: "WeatherAgent",
  handlers: {
    run: restate.createServiceHandler({ input: schema(WeatherPromptSchema), output: schema(z.string()) }, run),
  },
});

restate.serve({ services: [agent] });
